### PR TITLE
Add paintshop to stations

### PR DESCRIPTION
**Pull request by WKFO**
Adds paintshop to faction homeworld stations and some high population stations. There you can change the colors and the painting pattern of your ship. Pricing is calculated from the surface area approximated by adding up the cross section areas from all sides.

https://user-images.githubusercontent.com/29180904/168373985-dfdf3b38-800a-4f08-bfaf-451ae653e9fd.mp4

### CODE REVIEW

**Comment by impaktor** [comment]
Nice! I'm thinking (as I always do), that this shouldn't be available on all stations. Maybe something with station size? Tech level? Homeworld? (The other tabs are sort of required to be there, because "Frontier" / "It's always been like that").
2022-02-13-094303_1600x900_scrot

**Comment by impaktor** [comment]
Mockup, from IRC:

**Comment by WKFO** [comment]
One can now pick colors by hand, so I removed the color randomizer button without giving it much thought. Maybe I should've left it in.

**Comment by WKFO** [comment]
paint_prices

The current (placeholder) painting price algorithm is hullMass * 2.5

[10.0, 'varada']
[12.5, 'xylophis']
[20.0, 'pumpkinseed_police']
[25.0, 'kanara']
[25.0, 'kanara_civ']
[25.0, 'pumpkinseed']
[32.5, 'wave']
[45.0, 'amphiesma']
[62.5, 'bowfin']
[67.5, 'molamola']
[70.0, 'sinonatrix']
[75.0, 'lunarshuttle']
[80.0, 'sinonatrix_police']
[125.0, 'natrix']
[200.0, 'skipjack']
[437.5, 'deneb']
[525.0, 'bluenose']
[950.0, 'molaramsayi']
[1125.0, 'nerodia']
[1150.0, 'venturestar']
[1250.0, 'storeria']
[1950.0, 'ac33']
[2125.0, 'lodos']
[2225.0, 'vatakara']
[2350.0, 'malabar']
[3450.0, 'dsminer']

**Comment by Keeper1st** [comment]
To avoid having two "ship" icons, perhaps using the "fountain_pen" icon would be better, in lieu of a paint sprayer for now. I once again suggest not having text that refers to the player as "sir" as the pilot (and player) may not be male.

**Comment by WKFO** [comment]
Ah, "sir" is only for males right? English is my secondary language - I will change that. 

I think I can just wait for the paint sprayer unless it takes months.

**Comment by nozmajner** [comment]
I'll make one as soon as my school stuff winds down. Hopefully im the weekend the latest.

**Comment by Keeper1st** [comment]
Actually I thought of a better idea. Why not have the paintshop be part of the ShipRepair tab? If your ship is undamaged and the station meets the threshold of having the paintshop, then after the "Your ship is in perfect working condition" line there can be another -- potentially variable scripts: "We offer painting services. For your (shipname) we would charge $x. How about it?" / "Since you're here, how about we give your (shipname) a makeover? It'll cost you only $x." / "For $x we can customise your ship's paintjob." This would be followed by an "Enter paint shop" button. The user would not be charged at this point; the price is just mentioned up front.

Some of the the greeting texts in the paintshop would be changed as the user has been greeted already, so lines could be like "Please use this tool to show us how you'd like it." / "What design would you like your ship to have?" Then the apply paint button can say, "Apply paintjob for $x" perhaps.

This idea solves the problem of requiring an extra tab icon, and solves the mystery about who is talking to you about the paintjob, as it'd be assumed to be the station mechanic we see on the first part of the ShipRepair tab.

**Comment by impaktor** [comment]
If using a tab, as @Keeper1st suggests, we might want to consider also adding a line in the Lobby "This station rocks a paintshop" (but better worded), as else, people might not notice we even have painshops in the game? Ship repair tab is not something I usually clicked when I played the game.

**Comment by nozmajner** [comment]
That was my first thought as well. Javing its separate tab aids discoverability a lot.

**Comment by nozmajner** [comment]
@WKFO How about this one?
image
[icons-airbrush.zip](https://github.com/pioneerspacesim/pioneer/files/8130735/icons-airbrush.zip)

**Comment by WKFO** [comment]
@nozmajner Nice!

**Comment by Keeper1st** [comment]
If it's going to remain a separate tab, then I think A) the default tab icon size needs to change or they'd appear on top of some of the center tabs (I change the icon size in my own installation anyway as the default size is enormous) and B) there should be a facegen for the detailer. What would the title of the person be? "Ship Detailer" perhaps?

Another no-icon possibility would be on stations which have a ship detailer (i.e. paintshop), it shows up on the bulletin board as So-and-so's Ship Detailing Service. But we have the icon now, so...

If you're editing the icons file, how about a change to the Major Imports and Major Exports icons, so instead of only a slight change in size from the Minor versions, they also have two arrows or two arrow shafts, to make them a little more distinctive, suggesting visually a larger quantity of stuff coming or going?

**Comment by Zireael07** [comment]
I'd prefer if paintjob was a separate thing, OR the repair tab be renamed so that people do not miss the fact that it has more things than just repair.

**Comment by Keeper1st** [comment]
Just installed this for the first time. Started a new game at Mars, didn't have a paint shop, so flew to Brasilia. They had a paint shop but the price is shown as $0 until I actually try to apply the paint and get the "You do not have enough money" line. Then the price shows as $550. So there's an issue somewhere in the initial price listing. Also, it probably should tell the player up front that they don't have enough money rather than let them not realize they won't be able to pay for it. Perhaps the price should list as "Insufficient funds" in yellow to be consistent.

**Comment by WKFO** [comment]
The pricing thing is a remnant from the time when I thought of charging more for pattern changes (then discarded the idea). I will change that. Thanks for pointing out!

**Comment by impaktor** [comment]
@WKFO what's the status of this? You've done everything you wanted to do with this? Ready for merge?

(I'd squash merge, I think, unless you want to clean up the commit history)

**Comment by WKFO** [comment]
No more edits are planned. I did everything I wanted. Should be ready.

(Squash merge sounds fine.)

**Comment by Web-eWorks** [comment]
Couple of bugs / usability issues I found playing around with the paintshop feature:
- Clicking on the left arrow skips pattern 3 for the pumpkinseed, but clicking the right arrow allows it to be selected.
- The font choice for the paintshop screen is very uniform and makes it difficult to differentiate high-importance information (price) from low-importance information (individual RGB components of a color). I would recommend moving the price to near the bottom buttons and increasing the font size for it specifically.
- The "Apply Changes" should clearly indicate that you are purchasing the paintjob (e.g. "Confirm Purchase"), and I'd ideally move the Reset Preview button to underneath the color display in a smaller size so it's not confused for part of the purchase flow.
- The frame height for the pattern switch buttons doesn't match the text at 1600x900; see the related picture.

image

Finally, I'll second the requests for the paint shop to be moved to the Ship Repair tab, as there's really no reason to have a paint shop tab visible if the station doesn't support the paint shop, and it's fairly easy to add a line to the lobby or ship repairs tab letting players know the option is available. This isn't a hard requirement to merge the PR, but I think it's more semantically appropriate to be a subtab of the repair screen, and we are starting to have too many top-level tabs in the station view.

**Comment by nozmajner** [comment]
Hmm, if it goes under the repair tab, then maybe I could update the icon, and the station could pick the wrench and driver icon when there's no paint shop, and could pick an icon that indicates that paint shop is available if it is.

**Comment by impaktor** [comment]
- I agree with @Web-eWorks  that if feels more natural in the repair tab.

- not sure what to do with tab icon. Simplest it to do nothing / not using the new spray can

- optionally, some indication in the lobby could be nice, but I'd like it to be a bit more than just "we have spray shop", some ideas:
  - similar to black market, each repair? spray? shop is run by someone: "come by Standford & Son's ship repair and spray shop" + flavors
  - similar to tech level: "our ship repair & maintenance service is accredited for level 3" (player will learn what "level N" means by checking it out when they see it).
  - similar to discussion (on dev forum) on manufacturer licensing, so a station is licensed to deal with a specific ship manufacturer, although, not sure how that would tie in with spray shop, that aught to be manufacturer independent.

**Comment by Web-eWorks** [comment]
> not sure what to do with tab icon. Simplest it to do nothing / not using the new spray can

The tab icon would most likely be best served in the repair tab interface itself as a button to go to the paintjob options; the paintjob interface would fit most naturally as a "fullscreen modal" interface over top of / replacing the repair interface when activated, similar to e.g. the missions window.

> similar to tech level: "our ship repair & maintenance service is accredited for level 3" (player will learn what "level N" means by checking it out when they see it).

A "we offer ship repair, maintenance, and cosmetic services at this location" could suffice I think - along with renaming the "repair" tab to "Ship Maintenance" or similar.

**Comment by WKFO** [comment]
I don't know what's the deal with the pattern selector. In my Pioneer, Pumpkinseed only has 1 pattern anyway. It can cylce through all 10 patterns for Sinonatrix without problem though.

File: data/pigui/modules/station-view/06-shipRepairs.lua [hunk 1]
```diff
@@ -76,8 +91,148 @@ local function round(num, numDecimalPlaces)
...
 end
 
+local function determinePaintshopAvailability()
+	local player = Game.player
+	local station = player:GetDockedWith()
+	
+	-- faction homeworlds always have paintshops
+	if Game.system.faction ~= nil and Game.system.faction.hasHomeworld and Game.system.faction.homeworld == station.path:GetSystemBody().parent.path then
+		return true
+	end
```
[the remark this answers is not in the document]
Comment (WKFO) [reply data/pigui/modules/station-view/06-shipRepairs.lua:101]: I don't think 3-5 stations is a lot

File: data/pigui/modules/station-view/06-shipRepairs.lua [hunk 2]
```diff
@@ -76,8 +91,148 @@ local function round(num, numDecimalPlaces)
...
+	newColor["secondary"]["b"] = previewColors[2].b/255
+	newColor["secondary"]["a"] = previewColors[2].a/255
+	
+	newColor["trim"] = {}
+	newColor["trim"]["r"] = previewColors[3].r/255
+	newColor["trim"]["g"] = previewColors[3].g/255
+	newColor["trim"]["b"] = previewColors[3].b/255
+	newColor["trim"]["a"] = previewColors[3].a/255
+	
+	return newColor
```
[the remark this answers is not in the document]
Comment (WKFO) [reply data/pigui/modules/station-view/06-shipRepairs.lua:156]: I will try this as soon as I can

File: data/pigui/modules/station-view/06-shipRepairs.lua [hunk 3]
```diff
@@ -76,8 +91,148 @@ local function round(num, numDecimalPlaces)
...
+local function resetPreview()
+	local player = Game.player
+	previewPattern = player.model.pattern
+	previewSkin = player:GetSkin()
+	previewColors = previewSkin:GetColors()
+	
+	-- convert to a color format that works with ui.colorEdit
+	previewColors[1] = Color(string.format("%x%x%x", previewColors[1]["r"], previewColors[1]["g"], previewColors[1]["b"]))
+	previewColors[2] = Color(string.format("%x%x%x", previewColors[2]["r"], previewColors[2]["g"], previewColors[2]["b"]))
+	previewColors[3] = Color(string.format("%x%x%x", previewColors[3]["r"], previewColors[3]["g"], previewColors[3]["b"]))
```
[the remark this answers is not in the document]
Comment (Web-eWorks) [reply data/pigui/modules/station-view/06-shipRepairs.lua:217]: I believe the alpha channel is ignored by both the model pattern code and the specific color edit widget used to select these colors, it's treated as a "pure" RGB color without an alpha channel.

File: data/pigui/modules/station-view/06-shipRepairs.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (WKFO) [reply data/pigui/modules/station-view/06-shipRepairs.lua:102]: I don't intend to "fix" this unless there is a strong disagreement, so I'll mark this conversation as resolved - I like the idea of paintshops not being available everywhere, but I don't want paintshops to be pseudo-easter eggs. I would rather have a station not having a paintshop be an inconvenience, and not a station having a paintshop be a treasure. Following this philosophy, having paintshops everywhere on the capital planet is normal IMHO.

[the remark this answers is not in the document]
Comment (Web-eWorks) [reply data/pigui/modules/station-view/06-shipRepairs.lua:102]: Yeah, stations/planets with populations in the millions should logically have at least one paintshop somewhere I think; I'd rather the paintshop feature be available nigh-everywhere and reduce its availability at a later date to add ship maintenance gameplay (e.g. component repair, ammunition/life support/consumables restock, etc) than to make it really hard to find out of the gate.

**Pull request merged**